lms/resources/lti_launch.py

```diff
@@ -23,6 +22,17 @@ def __init__(self, request):
         self._request = request
         self._authority = self._request.registry.settings["h_authority"]
 
+    def get_or_create_course(self):
+        params = self._request.parsed_params
+        course = self._request.find_service(name="course").get_or_create(
+            self.h_group.authority_provided_id,
+            params["context_id"],
+            params["context_title"],
+            extra=self.course_extra,
+        )
+
+        return course
```

Comment:
I think we probably want this to be a service method on either `CourseService` or `GroupingService`, rather than an `LTILaunchResource` method.

The problem with `LTILaunchResource` is that it's only available during requests that use `LTILaunchResource` as the `context` (i.e. LTI launch requests). It's not available for example during API requests made by the backend to the frontend. Services are available to all requests though.

Reply:
This is get_or_create_course and just calls course_service.get_or_create with the right argument  (during a lti_launch) I think it could stay here but I'll add a docstring to see if it make sense.